src/main/java/net/imagej/ops/OpSignature.java

```diff
@@ -0,0 +1,188 @@
+
+package net.imagej.ops;
+
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.scijava.module.DefaultMutableModuleItem;
+import org.scijava.module.ModuleInfo;
+import org.scijava.module.ModuleItem;
+
+import net.imglib2.IterableInterval;
+import net.imglib2.RandomAccessible;
+import net.imglib2.RealRandomAccessible;
+import net.imglib2.type.numeric.RealType;
+
+/**
+ * The minimum required components for a call to
+ * {@link OpService#run(String, Object...)}
+ *
+ * @author Gabriel Selzer
+ */
+public class OpSignature {
+
+	private final String name;
+	private final String functionalType;
+	private final List<ModuleItem<?>> inputs;
+	private final List<ModuleItem<?>> outputs;
+
+	public OpSignature(final OpInfo info) {
+		this(info.getName(), info.getType(), info.inputs(), info.outputs());
+	}
+
+	public OpSignature(final String name, final Class<?> functionalType,
+		final List<ModuleItem<?>> inputs, final List<ModuleItem<?>> outputs)
+	{
+		this.name = name;
+		this.functionalType = functionalType.getName();
+		this.inputs = inputs;
+		this.outputs = outputs;
+	}
+
+	public String getName() {
+		return name;
+	}
+
+	public String getFunctionalType() {
+		return functionalType;
+	}
+
+	public List<ModuleItem<?>> inputsFor(ModuleInfo info) {
+		return inputs.stream() //
+			.map(item -> new DefaultMutableModuleItem<>(info, item)) //
+			.collect(Collectors.toList());
+	}
+
+	public List<Type> inputTypes() {
+		return inputs.stream() //
+			.map(ModuleItem::getGenericType) //
+			.collect(Collectors.toList());
+	}
+
+	public List<String> inputNames() {
+		return inputs.stream() //
+			.map(ModuleItem::getName) //
+			.collect(Collectors.toList());
+	}
+
+	public List<ModuleItem<?>> outputsFor(ModuleInfo info) {
+		return outputs.stream() //
+			.map(item -> new DefaultMutableModuleItem<>(info, item)) //
+			.collect(Collectors.toList());
+	}
+
+	public List<Type> outputTypes() {
+		return outputs.stream() //
+			.map(ModuleItem::getGenericType) //
+			.collect(Collectors.toList());
+	}
+
+	public List<String> outputNames() {
+		return outputs.stream() //
+			.map(ModuleItem::getName) //
+			.collect(Collectors.toList());
+	}
+
+	@Override
+	public boolean equals(Object that) {
+		// ensure that is a OpSignatureInfo
+		if (!(that instanceof OpSignature)) return false;
+		OpSignature thatSig = (OpSignature) that;
+		// check name
+		return buildShortSig().equals(thatSig.buildShortSig());
```

Comment:
I feel like this is not enough?

We should check name, functional type, input type and name, and output type and name.

Reply:
Fixed.